temporalio/converter.py

```diff
@@ -0,0 +1,362 @@
+"""Base converter and default implementations for conversion to/from values/payloads."""
+
+import dataclasses
+import inspect
+import json
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Tuple, Type
+
+import google.protobuf.json_format
+import google.protobuf.message
+import google.protobuf.symbol_database
+
+import temporalio.api.common.v1
+
+
+class DataConverter(ABC):
+    """Base converter to/from multiple payloads/values."""
+
+    @abstractmethod
+    async def encode(self, values: list[Any]) -> list[temporalio.api.common.v1.Payload]:
+        """Encode values into payloads.
+
+        Args:
+            values: Values to be converted.
+
+        Returns:
+            Converted payloads. Note, this does not have to be the same number
+            as values given, but at least one must be present.
+
+        Raises:
+            Exception: Any issue during conversion.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    async def decode(
+        self,
+        payloads: list[temporalio.api.common.v1.Payload],
+        type_hints: Optional[list[Type]],
+    ) -> list[Any]:
+        """Decode payloads into values.
+
+        Args:
+            payloads: Payloads to convert to Python values.
+            type_hints: Types that are expected if any. This may not have any
+                types if there are no annotations on the target. If this is
+                present, it must have the exact same length as payloads even if
+                the values are just "object".
+
+        Return:
+            Collection of Python values. Note, this does not have to be the same
+            number as values given, but at least one must be present.
+
+        Raises:
+            Exception: Any issue during conversion.
+        """
+        raise NotImplementedError
+
+
+class PayloadConverter(ABC):
+    """Base converter to/from single payload/value."""
+
+    @abstractmethod
+    async def encode(self, value: Any) -> Optional[temporalio.api.common.v1.Payload]:
```

Comment:
Not convinced this needs to be an async method, it shouldn't do any IO or CPU intensive operations.
The DataConverter should be async because it's responsible for compression and encryption.

Reply:
Y'all chose to make it async-capable in TypeScript, why not here? What are the concerns with making it async?

Reply:
I think it needs to be async unless we plan to separate out bytes-only stuff soon, as that easily could be cpu heavy for encryption (or need to do IO by reading keys from some source, etc)

Reply:
We're going to make the TS PayloadConverter synchronous.
In TS, async framework code could change the order of workflow execution which could break determinism but that might not apply here, and we'd need to restructure things a bit for that.
I would suggest that users don't do any CPU or IO in the `PayloadConverter` and instead do it in the higher level `DataConverter` where multiple payloads could be compressed together.

Reply:
> We're going to make the TS PayloadConverter synchronous.

Why? What are the concerns with making it async?

> In TS, async framework code could change the order of workflow execution which could break determinism but that might not apply here, and we'd need to restructure things a bit for that.

Why is this a problem for payload converters but not data converters?

>  I would suggest that users don't do any CPU or IO in the PayloadConverter and instead do it in the higher level DataConverter where multiple payloads could be compressed together.

While we might suggest this, the question is why is it ok for data converters to be async but not payload converters?

Reply:
In TS we're going to separate conversion into 2 steps, one sync that maps `object -> payload` and runs in workflow context, and the other async that maps `payloads -> payloads` and runs outside of workflow context.

Reply:
I see. I think if we were using subprocess sandboxing or similar we might need to do the same (and we'd also have to take the class of a data converter instead of an instance of it), but I am hoping our sandboxing does not cross some boundary requiring serialization to occur on the workflow side of that boundary. Though a method on a dataclass could provide a way to escape the sandbox. Will think on it.

Even if it does, I think the async payload conversion in the asyncio event loop will be in a predictable order.

Reply:
Here I don't expect the data converter to be inline with the workflow wrt determinism, nor am I more concerned about this change of a data converter mid-use any more than other types of data converter changes mid-use.